File: CE303_Assignment/src/assignment/Bank.java
package assignment;

import java.util.HashMap;

public class Bank {
    final HashMap<Integer, Trader> traders = new HashMap<>();

    public void createTrader(int traderId, boolean hasStock){
        Trader trader = new Trader(traderId, hasStock);
        addToTraders(trader);
    }

    public boolean hasTraderStock(int traderId) {
        return traders.get(traderId).hasStock();
    }

    public void transfer(int fromTrader, int toTrader) throws Exception {
        synchronized (traders) {
            if (!traders.get(fromTrader).hasStock())
                throw new Exception("Trader " + fromTrader + " does not have a stock. Unable to transfer.");
            traders.get(fromTrader).setStock(false);
            traders.get(toTrader).setStock(true);
        }
    }

    public HashMap<Integer, Trader> getCurrentTraders() {
        return traders;
    }

    public void addToTraders(Trader t) {
        traders.put(t.getTraderId(), t);
    }
}

File: CE303_Assignment_CS/BankServer/Bank.cs
using System;
using System.Collections.Generic;

namespace BankServer {
    class Bank {
        private readonly Dictionary<int, Trader> traders = new Dictionary<int, Trader> ();

        public void CreateTrader (int traderId, bool hasStock) {
            Trader trader = new Trader (traderId, hasStock);
            AddToTraders (trader);
        }

        public bool HasTraderStock (int traderId) {
            return traders[traderId].Stock;
        }

        public void Transfer (int fromTrader, int toTrader) {
            lock (traders) {
                if (!traders[fromTrader].Stock)
                    throw new Exception ($"Trader {fromTrader} + does not have a stock. Unable to transfer.");
                traders[fromTrader].Stock = false;
                traders[toTrader].Stock = true;

            }
        }

        public Dictionary<int, Trader> GetCurrentTraders () {
            return traders;
        }

        public void AddToTraders (Trader t) {
            traders.Add (t.TraderId, t);
        }

        public void RemoveFromTraders (int traderId) {
            traders.Remove (traderId);
        }
    }
}